### ISSUE TITLE
Auto reply on subscription

### ISSUE THREAD

**Issue by rezaa1** [knadh/listmonk#206]
Hi,
Is there any way we can send custom email after someone subscribed to a list. 

It can be a campaign  which remain active and send email as soon as someone subscribe to a list.

**Comment by jackdanielux** [comment]
If they're added to an opt in list (via API) post subscription they'll
automatically receive an opt in confirmation mail.

**Comment by rezaa1** [comment]
Yes, but I can not customize the opt in message for each list. can I?

**Comment by knadh** [comment]
listmonk doesn't have auto-responses yet. Will explore this.

**Comment by apkd** [comment]
Dropping by to let you know that I'm also looking forward to this feature!

**Comment by josephfugata** [comment]
Please. Looking forward to this as well

**Comment by ccfiel** [comment]
Any update on this?

**Comment by ccfiel** [comment]
Any work around?

**Comment by knadh** [comment]
Apologies for the delayed reply. There is no simple workaround for this unfortunately. A really clunky way would be to run a cron job every 10 mins that checks the subscription status + updated time of subscriptions in the `subscriber_lists` table and then triggers a transactional mail via the `/api/tx` API.

**Comment by t3hmrman** [comment]
Would the idea of adding a best-effort webhook firing at certain events (subscriber addition, cancellation, etc) be possible/likely to be merged if contributed?

I could see writing a sidecar to run alongside listmonk and watch the subscribers and make this change, but it would be *much* easier to write that sidecar if it was just a server waiting to receive a webhook as well!

**Comment by knadh** [comment]
This isn't trivial I feel:

- Have a queue into which events are buffered.
- Each event has to carry full medata (eg: a full list object or a subscriber object) which may not be currently available in certain API calls. For example, deleting a list doesn't fetch the list currently.
- UI in admin->settings to configure one or more webhook endpoints.
- Wait/max_retry mechanism for webhook HTTP requests.

**Comment by t3hmrman** [comment]
I definitely agree that it isn't trivial, but maybe a v1 could be? If the internal protocol is left quite bare you might be able to get away with:

- No queuing or buffering -- just a "best effort" web request. Maybe a background job or goroutine that can fail is called (best effort, if it fails, then it fails and you see a log message)
- Not every event has to be supported for hooks *at first* -- and the data exposed through the webhook does not necessarily have to match the full data of the action at first (I think!)
- UI is definitely hard requirement, with ENV for each supported event (something like `EVENT_<event>_WEBHOOK_URL`?)
- No retries? Just one chance and that's it?

I think if the protocol is at least in place, then as retries and other stuff get added it can be migration-free for downstream users (as the complexity behind the `Send()` method increases))

**Comment by candidexmedia** [comment]
@t3hmrman : have you (or anyone) found a workaround?

There is some interest in this feature on Twitter: https://twitter.com/statsandr/status/1625172227153854466?s=20

**Comment by t3hmrman** [comment]
Nope I haven't found a work around, but I haven't made a PR either -- if anyone wants to contribute a possible v1 they're free to!

**Comment by anaclumos** [comment]
I would also like to know if this is possible!

**Comment by Twix53791** [comment]
I confirm this feature is a _must_. I don't understand why it is so complicated to implement: the opt-in double option send actually emails automatically no? So why the opt-in simple process does not just follow the same process, sending an email (but a welcome email) and auto confirm the subscription behind?

**Comment by knadh** [comment]
@Twix53791 Auto-reply isn't complex but it isn't simple either. I just haven't had the time to prioritise it over other higher priority features yet, unfortunately.

- Have a new auto-reply templates section? Need the full e-mail body composing UI for this.
- Select different auto-responses for different lists (because the same auto-response will not make sense for different kinds of lists).

**Comment by ffabbr** [comment]
Any updates on this? Can we help you somehow?

**Comment by nielsmouthaan** [comment]
Ooh, not having this unfortunately prevents me from using Listmonk. Other than that, it looks great. I need it to send a series of onboarding emails. Heck, also just one email will do. It could be as simple as setting a (non-expiring) campaign as on "welcome campaign" setting on a list. I'll watch this issue and switch immediately from MailerLite to this once this gets in.

**Comment by knadh** [comment]
@nielsmouthaan would transactional e-mails suit your usecase (if you're triggering onboarding e-mails via APIs).

**Comment by peterjcarroll** [comment]
If there was a way to receive a notification of a new subscriber via a webhook, that would be helpful. Then at least I could use one of the many no-code workflow tools out there to trigger a welcome email.

**Comment by nielsmouthaan** [comment]
Exactly. Transactional email isn't the issue. It's the triggering (of the subscriber being added) that lacks.

**Comment by cloudroaminghub** [comment]
mark, need this great feature

**Comment by 1avender** [comment]
> Apologies for the delayed reply. There is no simple workaround for this unfortunately. A really clunky way would be to run a cron job every 10 mins that checks the subscription status + updated time of subscriptions in the `subscriber_lists` table and then triggers a transactional mail via the `/api/tx` API.

is there any such kind of workaround?

**Comment by cloudroaminghub** [comment]
> > Apologies for the delayed reply. There is no simple workaround for this unfortunately. A really clunky way would be to run a cron job every 10 mins that checks the subscription status + updated time of subscriptions in the `subscriber_lists` table and then triggers a transactional mail via the `/api/tx` API.
> 
> is there any such kind of workaround?


Here is a simple workaround script in centos, should add a 10 minutes job to crontab

#!/bin/bash

# Get the current time and format it as ISO 8601
current_time=$(date -u +"%Y-%m-%dT%H:%M:%SZ")

# Define a threshold (e.g., 10 minutes)
time_threshold=$(date -u -d "10 minutes ago" +"%Y-%m-%dT%H:%M:%SZ")

# Authentication credentials (replace with your own credentials)
tx_auth_credentials="username@password"

# Construct query parameters, including a check on created_at time range
query_params="page=1&per_page=100&query=subscribers.created_at%20>%20'$time_threshold'"

# Call the API to get the current list of subscribers
response=$(curl -u "$tx_auth_credentials" -X GET "http://localhost:9000/api/subscribers?$query_params")

# Use jq to parse the response
new_subscribers=$(echo $response | jq -r '.data.results[] | .email')

# Loop through new subscribers and send emails
for email in $new_subscribers
do
  # JSON data template
  json_data=$(cat <<EOF
{
    "subscriber_email": "$email",
    "template_id": 3,
    "data": {"order_id": "1234", "date": "2022-07-30", "items": [1, 2, 3]},
    "content_type": "html"
}
EOF
  )

  # Send the email
  curl -u "$tx_auth_credentials" "http://localhost:9000/api/tx" -X POST \
       -H 'Content-Type: application/json; charset=utf-8' \
       --data-binary "$json_data"
done

**Comment by cloudroaminghub** [comment]
another version to check subscription status to send different mail

```
#!/bin/bash

# Function to log messages
log_message() {
  echo "$(date -u +"%Y-%m-%dT%H:%M:%SZ") - $1"
}

# Get the current time in ISO 8601 format
current_time=$(date -u +"%Y-%m-%dT%H:%M:%SZ")
log_message "Current Time: $current_time"

# Define a time threshold (e.g., 10 minutes ago)
time_threshold=$(date -u -d "10 minutes ago" +"%Y-%m-%dT%H:%M:%SZ")
log_message "Time Threshold: $time_threshold"

# Authentication credentials from environment variables
tx_auth_credentials="USERNAME:PASSWORD"

# Construct query parameters
query_params="page=1&per_page=100&query=subscribers.created_at%20>%20'$time_threshold'"

# Function to send email based on subscription status
send_email() {
  local email=$1
  local subscription_status=$2
  local template_id

  # Determine the template based on subscription status
  case $subscription_status in
    "unconfirmed")
      template_id=3  # Example template ID for unconfirmed
      ;;
    "confirmed")
      template_id=4  # Example template ID for confirmed
      ;;
    *)
      template_id=5  # Default template ID for other cases
      ;;
  esac

  local json_data=$(cat <<EOF
{
  "subscriber_email": "$email",
  "template_id": $template_id,
  "data": {"order_id": "1234", "date": "2022-07-30", "items": [1, 2, 3]},
  "content_type": "html"
}
EOF
  )
  curl -u "$tx_auth_credentials" "http://localhost:9000/api/tx" -X POST \
       -H 'Content-Type: application/json; charset=utf-8' \
       --data-binary "$json_data"
  log_message "Email sent to $email with template $template_id"
}

# Call the API to get the current list of subscribers
response=$(curl -u "$tx_auth_credentials" -X GET "http://localhost:9000/api/subscribers?$query_params")
if [ $? -ne 0 ]; then
  log_message "Error fetching subscribers"
  exit 1
fi

# Use jq to parse the response and process each subscriber
echo $response | jq -c '.data.results[]' | while read -r subscriber; do
  email=$(echo $subscriber | jq -r '.email')
  subscription_status=$(echo $subscriber | jq -r '.lists[0].subscription_status')

  # Send email based on subscription status
  send_email "$email" "$subscription_status"
done
```

**Comment by 1avender** [comment]
put the temp solution [here](https://github.com/cloudroaminghub/listmonk-autoreply-onsubscription)

**Comment by cloudroaminghub** [comment]
put the temp solution [here](https://github.com/cloudroaminghub/listmonk-autoreply-onsubscription)

**Comment by poblabs** [comment]
Any updates here? My use case is to send subscribers a discount coupon once they confirm their subscription. Having a 10 minute delay may be a turn off for some potential customers. I don't want to run a cron minutely either for the low volume. Having some sort of webhook, event, or maybe even MQTT pub/sub action trigger once a subscription is confirmed would be beneficial

**Comment by ffabbr** [comment]
I agree, having some sort of email automation is a must-have. It can be the most basic and developer-first approach; but still.

**Comment by poblabs** [comment]
@knadh Any consideration for this with your 3.0.0 release? #1699 

Even if it's developer-forward, a way to automate new subscribers, and unsubscribers seems to be in high demand here.

**Comment by joerosenthal** [comment]
There's already a mechanism for sending an opt-in email—why not hijack that pathway to send a welcome email (an opt-in email with no opt-in button) and just mark the user as opted in automatically?

**Comment by ghostdevv** [comment]
I had to implement something like this for my clients site, for us it was to send a webhook to [n8n](https://n8n.io/) so we can send a series of marketing emails after someone joins a list. (also had to various other patches based on the `/api/tx` route to allow sending marketing emails without an attached campaign)

> There's already a mechanism for sending an opt-in email—why not hijack that pathway to send a welcome email (an opt-in email with no opt-in button) and just mark the user as opted in automatically?

My implementation was based on this idea, but there are a few api endpoints which don't call this optin system so that's worth considering.

**Comment by poblabs** [comment]
This is interesting and I'd like to learn more. Currently I am experimenting with a python script that is monitoring nginx proxy logs and any  POST to my list id is captured and the subscriber ID is extracted from the URL so it gets users info. Then it calls the `/campaign/test` api on a completed campaign attached to a 0 user list. With the payload body containing the email contents, template id, etc. A concern I have is the python script exiting without notice. So some way to get this from the core would be ideal. I have no Go experience which is why I was playing around with python

**Comment by ghostdevv** [comment]
@poblabs You can look at the code changes I made [here](https://github.com/ghostdevv/listmonk-tweaked), I also have no go experience so I stumbled my way through it xD

The webhook specific changes I made are:
- https://github.com/ghostdevv/listmonk-tweaked/commit/63585480f786983da7a9d28bf48be2fafaff009c
- https://github.com/ghostdevv/listmonk-tweaked/commit/04bfa65e9d9de168cdb11a4ade9da3e93b74e099

I've added a README to explain the changes I made in my fork, feel free to use/change it but I won't be supporting it beyond fixing bugs in the features we need until such a time that listmonk supports what we're after or a suitable alternative is found.

**Comment by candidexmedia** [comment]
Related, for drip campaigns: https://github.com/knadh/listmonk/issues/815

**Comment by witten** [comment]
I've got yet another work-around for this issue:

 * First, I [overrode Listmonk's static files](https://yasoob.me/posts/setting-up-listmonk-opensource-newsletter-mailing/#custom-static-files). This involves downloading a copy of Listmonk's static directory (which in my case I do with Ansible's `get_url`, grabbing a tarball from the [Listmonk releases pages](https://github.com/knadh/listmonk/releases), pointing Listmonk at it, and modifying one of the static files with the following hack.
 * I edited the `optin.html` template and added something like this:
```
{{ if eq .Data.Title "Confirmed" }}
    Here is the welcome content (or a link to it elsewhere) ...
{{ end }}
```
This means that that content is displayed only on the subscription confirmation page when a subscription is actually confirmed.

To be clear, this approach has a bunch of downsides:

 * The welcome email isn't in email! It's displayed on a web page.
 * It's somewhat ephemeral by its nature (it isn't in the user's email inbox), and so they can't return to it later for reference.
 * You can mitigate that somewhat by linking to content instead of including it inline in the page, but the user would still have to remember to bookmark it.
 * Any content you link to isn't password-protected or otherwise gated from a user sharing the URL. However if you keep the content inline, it isn't linkable.
 
The upsides are:

 * It's instant. The user doesn't have to wait for a ten-minute or even one-minute API poll cycle to complete for a welcome email to come in.
* You aren't at risk of double emailing welcome emails or missing sending of welcome emails if your API polling isn't perfect.
 * You don't have to use a modified version of Listmonk. (I'm not counting the custom static directory.)

I hope this is useful to someone! And I'll join the chorus of folks clamoring for welcome email functionality to be built into Listmonk.

**Comment by witten** [comment]
An update to the approach above. Here's what I'm doing now:

 * Instead of modifying `optin.html` from the static files, I'm altering `message.html`. That way, I can display the custom info *after* the subscription is confirmed rather than before.. And so someone has to actually be subscribed to get the welcome content.
 * I'm still doing the `{{ if eq .Data.Title "Confirmed" }}` trick described above, so it only displays upon subscription confirmation and not for other messages. I'm also—now hear me out—including a bit of JavaScript that parses the subscriber UUID out of the current URL. Why? Well, the subscriber UUID isn't available in the `message.html` template context, and I need it so that I can put a link on the `message.html` page that goes to...
 * So, I created a draft campaign that has the welcome content. (It also works if the campaign is status "finished," but then you can't edit it if you want to make changes.) The reason I'm using a campaign for this is so that I can just use the built-in Listmonk content management UI to edit and manage this content. It's also nice to be able to email it out to any current list subscribers if you want.
 * So, getting back to that `message.html` link. Courtesy of the aforementioned JavaScript, the link goes to the campaign "view in browser" page, the URL of which is constructed on the fly in the `message.html` JavaScript to go to the current subscriber's version of that campaign page.

Now, yes, this is total hack town. But it works pretty well! The user subscribes to the list. Once they confirm, they get a call to action link to read the welcome content they get for subscribing. They click the link and it takes them to their user-specific "view in browser" page for the welcome campaign. Would I rather have built-in welcome email functionality? Hell yeah. But for now, this appears to work.

**Comment by sspaeti** [comment]
@witten This sounds amazing and not too hacky, I'd like to try it. Thanks for taking the time to write it up. Would you mind sharing the Listmonk sign-up website, where I can see how that feels (so I know if I should start hacking or search for a workaround with other solutions)?

**Comment by witten** [comment]
The newsletter sign-up flow starts here: https://danhelfmanbooks.com/newsletter/

I'd be happy to delete your account after you try it out. Or I suppose you can just unsubscribe.

**Comment by pankajsoni19** [comment]
+1, can help switchover completely from AC

**Comment by witten** [comment]
IMO there's still a big need for a feature like this. (Replying so Stale Bot doesn't auto-close the ticket.)

**Comment by joerosenthal** [comment]
Seconded. I'm still using the first version of this software in production, for which I cobbled together a basic auto reply on subscription functionality. I'd love to upgrade this properly.

**Comment by iamcalledrob** [comment]
Echoing what others have said:

I'd LOVE to be using Listmonk, but basic drip campaigns are a requirement in my situation.

Crossing my fingers that this gets prioritized at some point, because otherwise Listmonk is about 1000x better than my current solution!

**Comment by ImaCrea** [comment]
Hi there, that would be a fantastic feature indeed. I guess in term of UX that could be set in the Campaigns section. With a type like "automated campaign" and user could select which action would trigger the campaign to be sent, like "on subscribe" "on unsubscribe"..

Ideally, triggers could be customised and scripted directly in the interface.

But first having a set of a few essentials triggers would be very helpful and pave the way to other suggestions.

Thank you for this great piece of software!

**Comment by jeaye** [comment]
Not stale.

**Comment by joerosenthal** [comment]
Coming back to this. We already send a welcome email for subscribers who need to opt-in. In the case of the welcome email, we would just send an email for non-opt-in users. This is effectively what I did when I hacked my version 1.0 that's still in production.

**Comment by jeaye** [comment]
Not stale.

**Comment by knadh** [comment]
Related (an underlying event/webhook system which is necessary for higher-level functionality like auto replies): https://github.com/knadh/listmonk/pull/2839

**Comment by joerosenthal** [comment]
I don't understand why we need an event/webhook system to send out welcome emails on sign up. We already have opt-in emails that go out if we require subscribers to opt in. We're essentially asking for an email to go out if someone signs up, even if an opt in isn't required.

**Comment by knadh** [comment]
I don't want to hardcode a specific "welcome e-mail" feature (UI/UX and logic in the backend). The extensible approach is to have an events system internally, which can then have an autoreply (or other actions like external webhooks) UX on the frontend. That gives the choice to send not just a welcome e-mail, but respond with custom templates to various scenarios.

- Signup
- Opt-in confirm
- Unsubscribe
- Blocklist etc.

**Comment by joerosenthal** [comment]
That makes sense to me. Hardcoding was is my less-than-ideal solution for my version 2.0 install that I've got in production. I can't really operate a mailing list without sending a welcome email.

**Comment by joerosenthal** [comment]
Just upgraded to version 6.10, and I've lost my hacked-together welcome message functionality from version 2. Any updates on this functionality being incorporated?

**Comment by sunjam** [comment]
What we need is equivalent of the test email, which can optionally be sent to the new subscriber.  That would resolve this: something as simple as...

Thank you for subscribing to
- `List you subscribed to`

[Click here](example.com) to Login.
[Click here](example.com) if you subscribed by mistake.

**Comment by karanrajpal** [comment]
Any chance this is being looked at? Looking forward to this functionality 🙏
